sympy/solvers/solveset.py

```diff
@@ -1008,7 +1012,44 @@ def _solveset(f, symbol, domain, _check=False):
                 result = _result - singularities
 
     if _check:
-        if isinstance(result, ConditionSet):
+        if (isinstance(result,ConditionSet)) and \
+                (domain.is_subset(S.Reals) and \
+                    (type(result) is not list)):
+            x = Symbol('x')
+            f = f.subs({symbol: x})
+            if _is_lambert(f,x):
+                if result.has(cos,sin):
+                    if (result.has(exp)) or (f.has(sqrt(x))) or\
+                        'Abs' in str(f):
+                        return result
+                    elif not domain.is_subset(S.Integers) and \
+                        ((result.has(cos) and not result.has(sin)) or \
+                            (result.has(sin) and not result.has(cos))) :
+                        f = solve(f,x)
+                        f = FiniteSet(*f)
+                        return f
+                elif domain.is_subset(S.Reals):
+                    if not 'sqrt' in str(f) and len(str(f)) < 120:
```

Comment:
Please don't use strings to do the test. You can do `if not any(i for i in f.atoms(Pow) if i.exp is S.Half) and f.count_ops() < smthng:`

Though I am not sure why there is a limit on the length/complexity of `f`.